ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala

```diff
@@ -50,66 +54,88 @@ object WebSocketService {
       self leftMap (e => f(e.shows))
   }
 
-  private final case class StepAndErrors[+LfV](
+  private final case class StepAndErrors[+Pos, +LfV](
       errors: Seq[ServerError],
-      step: InsertDeleteStep[domain.ActiveContract[LfV]]) {
+      step: InsertDeleteStep[(domain.ActiveContract[LfV], Pos)]) {
     import json.JsonProtocol._, spray.json._
-    def render(implicit lfv: LfV <~< JsValue): JsValue = {
-      def inj[V: JsonWriter](ctor: String) = (v: V) => JsObject(ctor -> v.toJson)
+    def render(implicit lfv: LfV <~< JsValue, pos: Pos <~< Map[String, JsValue]): JsValue = {
+      def inj[V: JsonWriter](ctor: String, v: V) = JsObject(ctor -> v.toJson)
+      val thisw =
+        Liskov.lift2[StepAndErrors, Pos, Map[String, JsValue], LfV, JsValue](pos, lfv)(this)
       JsArray(
-        step.deletes.iterator.map(inj("archived")).toVector
-          ++ Liskov.co[StepAndErrors, LfV, JsValue](lfv)(this).step.inserts.map(inj("created"))
-          ++ errors.map(inj[String]("error") compose (_.message)))
+        step.deletes.iterator.map(inj("archived", _)).toVector
+          ++ thisw.step.inserts.map {
+            case (ac, pos) =>
+              val acj = inj("created", ac)
+              acj copy (fields = acj.fields ++ pos)
+          } ++ errors.map(e => inj("error", e.message)))
```

Comment:
why don't we want to emit `errors` first? the client may decide to cancel the stream processing entirely based on what error(s) they receive.

Reply:
@hurryabit ^^^

Reply:
This block assembles a single message; the client will get all this at once.